src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2007,16 +2013,62 @@ protected bool MarkDefaultConstructor (TypeDefinition type, in DependencyInfo re
 			return MarkMethodIf (type.Methods, MethodDefinitionExtensions.IsDefaultConstructor, reason, sourceLocationMember) != null;
 		}
 
-		void MarkGetInstanceMethod (TypeDefinition type, in DependencyInfo reason, IMemberDefinition sourceLocationMember)
+		void MarkCustomMarshalerGetInstance (TypeDefinition type, in DependencyInfo reason, IMemberDefinition sourceLocationMember)
 		{
-			if (type?.HasMethods != true)
+			if (!type.HasMethods)
 				return;
 
 			MarkMethodIf (type.Methods, m =>
-				m.Name == "GetInstance" && m.Parameters.Count == 1 && m.Parameters[0].ParameterType.MetadataType == MetadataType.String,
+				m.Name == "GetInstance" && m.IsStatic && m.Parameters.Count == 1 && m.Parameters[0].ParameterType.MetadataType == MetadataType.String,
 				reason, sourceLocationMember);
 		}
 
+		void MarkICustomMarshalerMethods (TypeDefinition type, in DependencyInfo reason, IMemberDefinition sourceLocationMember)
+		{
+			do {
+				if (!type.HasInterfaces)
+					continue;
+
+				foreach (var iface in type.Interfaces) {
+					var iface_type = iface.InterfaceType;
+					if (!iface_type.IsTypeOf ("System.Runtime.InteropServices", "ICustomMarshaler"))
+						continue;
+
+					//
+					// Instead of trying to guess where to find the interface declaration linker walks
+					// the list of implemented interfaces and resolve the declaration from there
+					//
+					var tdef = iface_type.Resolve ();
+					if (tdef == null) {
+						HandleUnresolvedType (iface_type);
+						return;
+					}
+
+					MarkMethodsIf (tdef.Methods, m => {
+						if (m.IsStatic)
+							return false;
+
+						if (m.Parameters.Count == 1) {
+							var ptype = m.Parameters[0].ParameterType.MetadataType;
+							return m.Name switch
+							{
+								"MarshalNativeToManaged" when ptype == MetadataType.IntPtr => true,
+								"MarshalManagedToNative" when ptype == MetadataType.Object => true,
+								"CleanUpNativeData" when ptype == MetadataType.IntPtr => true,
+								"CleanUpManagedData" when ptype == MetadataType.Object => true,
+								_ => false
+							};
+						}
+
+						return m.Name == "GetNativeDataSize" && !m.HasParameters;
+					}, reason, sourceLocationMember);
```

Comment:
Is there a scenario where this extra code would be useful? I usually go by "less is better" philosophy.

```suggestion
					MarkMethodsIf (tdef.Methods, m => !m.IsStatic, reason, sourceLocationMember);
```